### ISSUE TITLE
Lambda in template context fails to implicitly capture const local variable (accepted by GCC/MSVC)

### ISSUE THREAD

**Issue by starbugs-qurong** [llvm/llvm-project#131452]
In a template context, when a lambda inside a function template uses a const-qualified local variable without explicit capture, Clang incorrectly rejects the code while GCC and MSVC accept it according to the C++ standard. If we replace the function template with a regular function then clang does not emit an error. 

For this program:

```cpp
template<typename T> 
void f() {
    const int i = 42;
    auto k = [] {  //  Error in Clang
        return i;  // Missing capture of 'i' in template context
    }();
}
```

This causes an error in clang, but MSVC and GCC accept it. 

Compiler Explorer link: https://godbolt.org/z/67T68WMhs

**Comment by frederick-vs-ja** [comment]
It seems that Clang mistakenly considers `i` to be odr-used (note that `i` is not odr-usable here). Replacing `return i;` with `return +i;` works ([Godbolt link](https://godbolt.org/z/5Tbb7T8j3)).

**Comment by shafik** [comment]
We have one issue w/ a lot of different cases, so I will use that one as the canonical dup.
